homework_2/internal/infrastructure/http/axilisOfferFeed.go

```diff
@@ -3,14 +3,18 @@ package http
 import (
 	"context"
 	"encoding/json"
+	"io/ioutil"
 	"log"
 	"net/http"
+	"strconv"
+	"strings"
 	"time"
 
 	"code-cadets-2021/lecture_2/06_offerfeed/internal/domain/models"
 )
 
 const axilisFeedURL = "http://18.193.121.232/axilis-feed"
+const axilisFeedURL2 = "http://18.193.121.232/axilis-feed-2"
```

Comment:
And why should you structure the code differently?

Maybe in a year there will be a new company - `InfinumFeeder`, which wont expose their feed through HTTP server.
They could expose their feed over gRPC, or some other protocol.

Then, you would already be prepared:
- you would create a new `infinumFeed.go` file which is an implementation of `Feed` interface
- thats it, just plug it in your code

Reply:
Thank you. Does it mean that I cannot use the same stuct axilisOfferOdd and axilisOfferOddDetails as in first feed, even though they would have the same fields? So I make new structs and later combine them into one channel?

Reply:
In the first feed, you use `axilisOfferOdd` and `axilisOfferOddDetails` structs just for parsing. (Only first feed should know about those structs because only first feed is contacting that endpoint).
First feed is immediately mapping `axilisOfferOdd` to application model `models.Odd`.

In the second feed you should not use the same structs. (And you are not using them, take a look at your `processResponse2` method). You implemented custom parsing and you are immediately mapping that strings & floats to application model `models.Odd`.

---
A little bit of theory:

Each feed defines its DTO models (data transfer object). For first feed we are using `axilisOfferOdd` and for second we are using some custom parsing (we do not define models at all).

Only feeds should know about different DTO's. The rest of the application should know only about domain models `models.Odd`.

The reason is:
- what if first feed changes its model? You want to update first feed only then, and the rest of the application should still work properly
- you achieve that by using domain models in the application, and separate DTO models for interaction with other services

Reply:
Thank you, I actually wanted to ask about AxilisOfferFeed struct but I typed in wrong. I understand now and I pushed my changes. I opted for the second option, I now have 2 AxilisOfferFeeds and a FeedMerger, in which I combined both channels into one. I did it by using new goroutines, where each reads from one channel and writes to the merge channel, and the main goroutine waits for them. I don't know if there is maybe a better way to do that.

Reply:
Np, it is looking vey good now.